notes/overriding.md

```diff
@@ -0,0 +1,676 @@
+
+There's more potential optimizations, but this first step should be a reasonable
+trade-off between implementation complexity and performance. See
+[#103](https://github.com/tweag/nickel/issues/103) for more details.
+
+### Priorities
+
+In some cases, the semantics of the proposed recursive annotations `default rec`
+and `force rec` are not obvious:
+
+```nickel
+let record = {
+  a = 1,
+  b | force = "DONT CHANGE",
+} in
+record | default rec
+```
+
+What should be the result? Several possibilities:
+
+1. The `default` erases everything else. This doesn't sound right, because a
+   configuration may use `force` to prevent a final value from being modified.
+2. The `default` only erases the `0` priority, and let any other untouched.
+3. The `default` erases everything excepted a `force`.
+
+Both 2. and 3. sounds reasonable. The `force rec` may have a dual behavior, but
+could also erase everything else: it make more sense to turn a `default` to a
+`force` than the other way around.
+
+### Custom merge functions
+
+In the case of the NixOS module system, all the configuration is merged in one
+final phase that conceptually evaluates the merge AST.
+
+In Nickel, merging can be done partially and observed by expressions. The
+natural semantics of custom merge functions should be to affect all subsequent
+merging, as for contracts application. Indeed, we want to specify a contract or
+a custom merge function once and for all, not to repeat it at every field
+definition:
+
+```nickel
+let add = fun args => args.lower + args.higher in
+// {a = 3}
+{a | merge add = 1} & {a = 1} & {a = 1}
+```
+
+However, this breaks commutativity, because it affects terms on the right but
+not on the left:
+
+```
+let r1 = {a = 1} in
+let r2 = {a = 1} in
+let r3 = {a | merge add = 1} in
+
+// {val = 2}
+r1 & r2 & r3
+
+// {val = 3}
+r3 & r1 & r2
+```
+
+Possible solutions:
+
+1. Make a `n-ary` merging behave differently than repeated merging, which would
+   apply the custom merge function to all arguments. This doesn't solve the
+   commutativity problem per se: we also have to decide that in ordinary
+   merging, custom function wouldn't affect subsequent merges anymore. This does
+   raise yet other problems, one being of a good distinctive syntax. More
+   generally, it looks confusing and unsatisfying.
+
+2. What seems the right, but not trivial solution, would be to symmetrize custom
+   merging and make it "distribute backward" as well, that is (simplifying by
+   forgetting about priority and `lower`/`higher` arguments):
+
+   ```nickel
+   val1 & val2 & (val3 | merge func)
+   <=> (val1 | merge func) & val2 & val3
+   <=> func (func val1 val2) val3
+   // instead of the naive
+   <=/=> func (val1 & val2) val3
+   ```
+
+   But what if `val1 & val2` is in a thunk that happens to be forced in between?
+   What should `let x = val1 & val2 in x & (val3 | merge func)` evaluate to? And
+   `let x = val1 & val2 in builtins.seq x (x & (val3 | merge func))`?
```

Comment:
I think forcing and “closing” a record should be too distinct operations.

The way I understand the rest of the semantics above, there's two kind of records: open and closed. Open records are the one available in the surface language and with all the shiny overriding semantics while closed records are an internal thing only.

If we represent closed records with `{| |}` (to distinguish them from open records `{ }`), we have that

1. `{ x = 1, y = x } == fun self => {| x= 1, y = self.y |}` (the `repr(r)` operator above)
2. `{| something |} == let fix = { something } fix in fix` (what we can call the `close(r)` operator)

To rephrase what you defined before, we then have that for an open record `r`, `r.foo == close(r).foo`.

All this to say that `val1 & val2` is still an open record (whether you force it or not), so `let x = val1 & val2 in builtins.seq x (x & (val3 | merge func))` is the same as `val1 & val2 & (val3 | merge func)`

Reply:
I was not even thinking about records, but values in general. The problem is, morally, that using a custom merge function amounts to interpret the merge AST with a different semantics for `&`. If we want it to be non-local, as in only one term with an annotation in the merge expression is sufficient to change the merge function for all operands, you need to have access to the whole original AST. Forcing thunks may evaluate sub-expressions with the original merge semantics (and not the custom, because it is not in the scope of the sub-expression), and this changes the result.

Example:

```
let add = fun x y => x + y in
let val | merge add = 1 in
// Say we do "the right thing", so this is 3 as expected
1 & 1 & val

let x = 1 & 1 in
// This has no chance of being 3
// x is forced, without any custom merge function in sight.
// `1 & 1` is thus computed to be `1` (merge is idempotent)
// The thunk of x is updated with 1, while the original expression `1 & 1` is lost.
// so x is `1` and the final result is `2`
seq x (x & val)
```
Merging numbers directly is dubious, but you can imagine them being values for an option: `{ ..., someOption | Num | merge add, ...}`.

The next paragraph notices that, though, records will need to be instrumented to store the original expression along the computed thunk (as a way of representing open records), and in that case we can make it work. Which is I think what you meant? But, in general, `val1` and others are not necessarily records nor record fields.